feat(remix): Enable Remix SDK (#5327)
getsentry/sentry-javascript: packages/remix/README.md

[![npm version](https://img.shields.io/npm/v/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![npm dm](https://img.shields.io/npm/dm/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![npm dt](https://img.shields.io/npm/dt/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![typedoc](https://img.shields.io/badge/docs-typedoc-blue.svg)](http://getsentry.github.io/sentry-javascript/)

This SDK is considering experimental and in an alpha state. It may experience breaking changes. Please reach out on [GitHub](https://github.com/getsentry/sentry-javascript/issues/new/choose) if you have any feedback/concerns.
## General

This package is a wrapper around `@sentry/node` for the server and `@sentry/react` for the client, with added functionality related to Remix.

To use this SDK, initialize Sentry in your Remix entry points for both the client and server.

```ts
// entry.client.tsx

import { useLocation, useMatches } from "@remix-run/react";
import * as Sentry from "@sentry/remix";
import { useEffect } from "react";

Sentry.init({
  dsn: "__DSN__",
  tracesSampleRate: 1,
  integrations: [
    new Sentry.BrowserTracing({

>         routingInstrumentation: Sentry.remixRouterInstrumentation(

Reviewer: nit: Why don't we set these properties for the user inside of `remixRouterInstrumentation` instead of asking them to do this?

Author: Because dependency injection means that we don’t have to deal with importing the functions ourselves, which means we minimize versioning conflicts.

also init is usually only done once, so it’s not like folks have to come back and touch this code again, so just letting them pass in the imports is fine.

we’ve done this for the rest of our routing instrumentation and never gotten negative feedback.

import { ErrorBoundary, withSentryRouteTracing } from "@sentry/remix";

function App() {
    return (
      <ErrorBoundary>
        <html>
          <head>
            <Meta />
            <Links />
          </head>
          <body>
            <Outlet />
            <ScrollRestoration />
            <Scripts />
            <LiveReload />
          </body>
        </html>
      </ErrorBoundary>
  );
}

export default withSentryRouteTracing(App);
```

To set context information or send manual events, use the exported functions of `@sentry/remix`.

```ts
import * as Sentry from '@sentry/remix';

// Set user information, as well as tags and further extras
Sentry.configureScope(scope => {
  scope.setExtra('battery', 0.7);
  scope.setTag('user_mode', 'admin');
  scope.setUser({ id: '4711' });
  // scope.clear();
});

// Add a breadcrumb for future events
Sentry.addBreadcrumb({
  message: 'My Breadcrumb',
  // ...
});

// Capture exceptions, messages or manual events
Sentry.captureMessage('Hello, world!');
Sentry.captureException(new Error('Good bye'));
Sentry.captureEvent({
  message: 'Manual',
  stacktrace: [
    // ...
  ],
});
```

## Sourcemaps and Releases

> The Remix SDK provides a script that automatically creates a release and uploads sourcemaps. To generate sourcemaps with Remix, you need to call `remix build` with `--sourcemap` option.

Reviewer 2: ```suggestion
The Remix SDK provides a script that automatically creates a release and uploads sourcemaps. To generate sourcemaps with Remix, you need to call `remix build` with the `--sourcemap` option.
```

# Official Sentry SDK for Remix

[![npm version](https://img.shields.io/npm/v/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![npm dm](https://img.shields.io/npm/dm/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![npm dt](https://img.shields.io/npm/dt/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/remix)
[![typedoc](https://img.shields.io/badge/docs-typedoc-blue.svg)](http://getsentry.github.io/sentry-javascript/)

This SDK is considering experimental and in an alpha state. It may experience breaking changes. Please reach out on [GitHub](https://github.com/getsentry/sentry-javascript/issues/new/choose) if you have any feedback/concerns.
## General

> This package is a wrapper around `@sentry/node` for the server and `@sentry/react` for the client, with added functionality related to Next.js.

Reviewer 3: ```suggestion
This package is a wrapper around `@sentry/node` for the server and `@sentry/react` for the client, with added functionality related to Remix.
```

> To use this SDK, init Sentry in your remix entry points for both the client and server.

Reviewer 3: ```suggestion
To use this SDK, init Sentry in your Remix entry points for both the client and server.
```

> ```javascript

Reviewer 3: ````suggestion
```typescript
````

// entry.client.tsx

import { useLocation, useMatches } from "@remix-run/react";
import * as Sentry from "@sentry/remix";
import { useEffect } from "react";

Sentry.init({
  dsn: "__DSN__",
  tracesSampleRate: 1,
  integrations: [
    new Sentry.BrowserTracing({
        routingInstrumentation: Sentry.remixRouterInstrumentation(
            useEffect,
            useLocation,
            useMatches,
        ),
    }),
  ],
  // ...
});
```

> 

Reviewer 4: ```suggestion

Also, wrap your Remix root, with `ErrorBoundary` to catch React errors, and `withSentryRouteTracing` to get parameterized router transactions.

```typescript
// root.tsx

import {
  Links,
  LiveReload,
  Meta,
  Outlet,
  Scripts
  ScrollRestoration,
} from "@remix-run/react";

import { ErrorBoundary, withSentryRouteTracing } from "@sentry/remix";

function App() {
    return (
      <ErrorBoundary>
        <html>
          <head>
            <Meta />
            <Links />
          </head>
          <body>
            <Outlet />
            <ScrollRestoration />
            <Scripts />
            <LiveReload />
          </body>
        </html>
      </ErrorBoundary>
  );
}

export default withSentryRouteTracing(App);

```

Reviewer 3: Not sure if that's the "Remix way" of doing thing tbh 🤔
https://remix.run/docs/en/v1/guides/constraints#higher-order-functions

I know the docs are only talking about `action`/`loader`, but I think it's meant to also reference the component
CC/ @kentcdodds

Author: Considering this wrapping happens only happens once for your entire app, we can get away with the single use of a HOC - especially since the alternative is some pretty unwieldy monkey patching

Reviewer 4: Hey @MichaelDeBoey 👋 

The challenge is to be able to get the parameterized routes for `pageload`/`navigation` transactions (they are labelled and grouped by routes in Sentry platform).

For React Router versions < 6, it's not an issue, as we can reach the `history` object, and match with the routes defined on initialization all together, without needing to wrap components.

Since React Router 6, we need hooks to get those, [so we ended up wrapping `<Routes />` component where those hooks are functional](https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/react-router/#react-router-v6).

It's similar here for Remix, except we don't have `<Routes />`, but wrapping `App` allows us to instrument all routes at once.

About the `ErrorBoundary`, we already discussed to make it a part of the final wrapper, but it's not yet implemented.

Any suggestions of alternative ways to reach parameterized routes would be great, and we can simplify both Remix and React Router 6 instrumentations.

Thanks!

> ```javascript

Reviewer 3: ````suggestion
```typescript
````

// entry.server.tsx

import { prisma } from "~/db.server";

import * as Sentry from "@sentry/remix";

Sentry.init({
  dsn: "__DSN__",
  tracesSampleRate: 1,
  integrations: [new Sentry.Integrations.Prisma({ client: prisma })],
  // ...
});
```

To set context information or send manual events, use the exported functions of `@sentry/remix`.

```javascript
import * as Sentry from '@sentry/remix';

// Set user information, as well as tags and further extras
Sentry.configureScope(scope => {
  scope.setExtra('battery', 0.7);
  scope.setTag('user_mode', 'admin');
  scope.setUser({ id: '4711' });
  // scope.clear();
});

// Add a breadcrumb for future events
Sentry.addBreadcrumb({
  message: 'My Breadcrumb',
  // ...
});

// Capture exceptions, messages or manual events
Sentry.captureMessage('Hello, world!');
Sentry.captureException(new Error('Good bye'));
Sentry.captureEvent({
  message: 'Manual',
  stacktrace: [
    // ...
  ],
});
```

## Sourcemaps and Releases

The Remix SDK provides a script that automatically creates a release and uploads sourcemaps.

> 

Reviewer 4: ```suggestion
To generate sourcemaps with Remix, you need to call `remix build` with `--sourcemap` option. 
```
